sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSessionManager.java

```diff
@@ -277,7 +277,13 @@ Mono<ServiceBusReceiveLink> getActiveLink() {
         }
         return Mono.defer(() -> createSessionReceiveLink()
             .flatMap(link -> link.getEndpointStates()
-                .takeUntil(e -> e == AmqpEndpointState.ACTIVE)
+                .filter(e -> e == AmqpEndpointState.ACTIVE)
+                .next()
+                // It is possible that link is CLOSED directly from UNINITIALIZED state. This may happen when remote
+                // refuses to attach link and the error condition in 'onLinkRemoteClose()' is null, so 'endpointStates'
+                // will emit the complete signal. In this case, we switch to Mono.error() in order to do the retry.
+                .switchIfEmpty(Mono.error(() ->
+                    new AmqpException(true, "Session receive link completed without being active", null)))
```

Comment:
thank you, this operator chain looks what we want!

Reply:
@anuchandy How often will this chain be called? I was thinking back to where we had that high memory usage because it was eagerly assembling and creating the exception.... Do we need to consider `Mono.defer(() -> Mono.error(...))`?

Reply:
Hi @conniey - 


Looking at the code, my understanding is - Initially, there is going to be a max-concurrent-session instantiation of this chain in one shot. After that, this chain will be instantiated each time there is a session disconnect. That looks like a relatively low object creation rate in typical cases, provided the broker won't disconnect that often. If not enough requested sessions are available in the broker, the corresponding sessions acquire retry attempts going to instantiate that many chains concurrently.

I think, in the past, that high memory consumption was when we were using "Mono.just(item)" and "Mono.error(new Error)". At that time we replaced "Mono.just(item)" with its delayed version, "Mono.defer(() -> just(item))". The PR [reference](https://github.com/Azure/azure-sdk-for-java/pull/26373).

In this case, we use "Mono.error(() -> new Error())" which is a delayed version of "Mono.error(new Error())". Which should be equivalent of what you mentioned "Mono.defer(() -> Mono.error(...))"

Reply:
Nice! Thank you for clarifying. :)